notebook/base/handlers.py

```diff
@@ -411,6 +412,39 @@ def check_xsrf_cookie(self):
             return
         return super(IPythonHandler, self).check_xsrf_cookie()
 
+    def check_host(self):
+        """Check the host header if remote access disallowed.
+
+        Returns True if the request should continue, False otherwise.
+        """
+        if self.settings.get('allow_remote_access', False):
+            return True
+
+        # Remove port (e.g. ':8888') from host
+        host = re.match(r'^(.*?)(:\d+)?$', self.request.host).group(1)
+
+        # Browsers format IPv6 addresses like [::1]; we need to remove the []
+        if host.startswith('[') and host.endswith(']'):
+            host = host[1:-1]
+
+        try:
+            addr = ipaddress.ip_address(host)
+        except ValueError:
+            # Not an IP address: check against hostnames
+            allow = host in self.settings.get('local_hostnames', [])
+        else:
+            allow = addr.is_loopback
+
+        if not allow:
+            self.log.warning("Blocking request with non-local 'Host' %s (%s)",
```

Comment:
Maybe add a reference to `NotebookApp.allow_remote_access` so people who see this message can take the appropriate action if they want to change it (e.g. localhost is proxied through another service)?